### Libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


### Get Data

In [64]:
train = pd.read_csv("../../data/train.csv")
test = pd.read_csv("../../data/test.csv")

### Data Exploration

In [65]:
n_train = train.shape[1]
print("The train dataset has %d features and %d records." % (train.shape[1]-1, train.shape[0]))

print("The train dataset has %d records." % test.shape[0])

The train dataset has 80 features and 1460 records.
The train dataset has 1459 records.


In [12]:
train.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

#### Data Pre-processing
<ol>
    <li>Separating the dependent variable <code>SalePrice</code> from the feature matrix.
    <li>Combining the <i>train</i> and <i>test</i> dataset.
</ol>

In [66]:
y_train = train['SalePrice']

dat = pd.concat([train.iloc[:,:-1], test])

print('The count of data types of the features of the combined dataset: \n', dat.dtypes.value_counts())

The count of data types of the features of the combined dataset: 
 object     43
int64      26
float64    11
dtype: int64


In [67]:
pd.set_option('display.max_columns', 100)
print('Summary Statistics for combined dataset columns with object datatype')
dat.describe(include = 'object')

Summary Statistics for combined dataset columns with object datatype


MSZoning Street Alley LotShape LandContour Utilities LotConfig  \
count      2915   2919   198     2919        2919      2917      2919   
unique        5      2     2        4           4         2         5   
top          RL   Pave  Grvl      Reg         Lvl    AllPub    Inside   
freq       2265   2907   120     1859        2622      2916      2133   

       LandSlope Neighborhood Condition1 Condition2 BldgType HouseStyle  \
count       2919         2919       2919       2919     2919       2919   
unique         3           25          9          8        5          8   
top          Gtl        NAmes       Norm       Norm     1Fam     1Story   
freq        2778          443       2511       2889     2425       1471   

       RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType ExterQual  \
count       2919     2919        2918        2918       2895      2919   
unique         6        8          15          16          4         4   
top        Gable  CompShg     VinylSd     VinylSd       None        TA   
freq        2310     2876        1025        1014       1742      1798   

       ExterCond Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  \
count       2919       2919     2838     2837         2837         2840   
unique         5          6        4        4            4            6   
top           TA      PConc       TA       TA           No          Unf   
freq        2538       1308     1283     2606         1904          851   

       BsmtFinType2 Heating HeatingQC CentralAir Electrical KitchenQual  \
count          2839    2919      2919       2919       2918        2918   
unique            6       6         5          2          5           4   
top             Unf    GasA        Ex          Y      SBrkr          TA   
freq           2493    2874      1493       2723       2671        1492   

       Functional FireplaceQu GarageType GarageFinish GarageQual GarageCond  \
count        2917        1499       2762         2760       2760       2760   
unique          7           5          6            3          5          5   
top           Typ          Gd     Attchd          Unf         TA         TA   
freq         2717         744       1723         1230       2604       2654   

       PavedDrive PoolQC  Fence MiscFeature SaleType SaleCondition  
count        2919     10    571         105     2918          2919  
unique          3      3      4           4        9             6  
top             Y     Gd  MnPrv        Shed       WD        Normal  
freq         2641      4    329          95     2525          2402

In [38]:
print('Summary Statistics for combined dataset columns with int64 datatype')
dat.describe(include = 'int64')

Summary Statistics for combined dataset columns with int64 datatype


Id   MSSubClass        LotArea  OverallQual  OverallCond  \
count  2919.000000  2919.000000    2919.000000  2919.000000  2919.000000   
mean   1460.000000    57.137718   10168.114080     6.089072     5.564577   
std     842.787043    42.517628    7886.996359     1.409947     1.113131   
min       1.000000    20.000000    1300.000000     1.000000     1.000000   
25%     730.500000    20.000000    7478.000000     5.000000     5.000000   
50%    1460.000000    50.000000    9453.000000     6.000000     5.000000   
75%    2189.500000    70.000000   11570.000000     7.000000     6.000000   
max    2919.000000   190.000000  215245.000000    10.000000     9.000000   

         YearBuilt  YearRemodAdd     1stFlrSF     2ndFlrSF  LowQualFinSF  \
count  2919.000000   2919.000000  2919.000000  2919.000000   2919.000000   
mean   1971.312778   1984.264474  1159.581706   336.483727      4.694416   
std      30.291442     20.894344   392.362079   428.701456     46.396825   
min    1872.000000   1950.000000   334.000000     0.000000      0.000000   
25%    1953.500000   1965.000000   876.000000     0.000000      0.000000   
50%    1973.000000   1993.000000  1082.000000     0.000000      0.000000   
75%    2001.000000   2004.000000  1387.500000   704.000000      0.000000   
max    2010.000000   2010.000000  5095.000000  2065.000000   1064.000000   

         GrLivArea     FullBath     HalfBath  BedroomAbvGr  KitchenAbvGr  \
count  2919.000000  2919.000000  2919.000000   2919.000000   2919.000000   
mean   1500.759849     1.568003     0.380267      2.860226      1.044536   
std     506.051045     0.552969     0.502872      0.822693      0.214462   
min     334.000000     0.000000     0.000000      0.000000      0.000000   
25%    1126.000000     1.000000     0.000000      2.000000      1.000000   
50%    1444.000000     2.000000     0.000000      3.000000      1.000000   
75%    1743.500000     2.000000     1.000000      3.000000      1.000000   
max    5642.000000     4.000000     2.000000      8.000000      3.000000   

       TotRmsAbvGrd   Fireplaces   WoodDeckSF  OpenPorchSF  EnclosedPorch  \
count   2919.000000  2919.000000  2919.000000  2919.000000    2919.000000   
mean       6.451524     0.597122    93.709832    47.486811      23.098321   
std        1.569379     0.646129   126.526589    67.575493      64.244246   
min        2.000000     0.000000     0.000000     0.000000       0.000000   
25%        5.000000     0.000000     0.000000     0.000000       0.000000   
50%        6.000000     1.000000     0.000000    26.000000       0.000000   
75%        7.000000     1.000000   168.000000    70.000000       0.000000   
max       15.000000     4.000000  1424.000000   742.000000    1012.000000   

         3SsnPorch  ScreenPorch     PoolArea       MiscVal       MoSold  \
count  2919.000000  2919.000000  2919.000000   2919.000000  2919.000000   
mean      2.602261    16.062350     2.251799     50.825968     6.213087   
std      25.188169    56.184365    35.663946    567.402211     2.714762   
min       0.000000     0.000000     0.000000      0.000000     1.000000   
25%       0.000000     0.000000     0.000000      0.000000     4.000000   
50%       0.000000     0.000000     0.000000      0.000000     6.000000   
75%       0.000000     0.000000     0.000000      0.000000     8.000000   
max     508.000000   576.000000   800.000000  17000.000000    12.000000   

            YrSold  
count  2919.000000  
mean   2007.792737  
std       1.314964  
min    2006.000000  
25%    2007.000000  
50%    2008.000000  
75%    2009.000000  
max    2010.000000

In [39]:
print('Summary Statistics for combined dataset columns with float64 datatype')
dat.describe(include = 'float64')

Summary Statistics for combined dataset columns with float64 datatype


LotFrontage   MasVnrArea   BsmtFinSF1   BsmtFinSF2    BsmtUnfSF  \
count  2433.000000  2896.000000  2918.000000  2918.000000  2918.000000   
mean     69.305795   102.201312   441.423235    49.582248   560.772104   
std      23.344905   179.334253   455.610826   169.205611   439.543659   
min      21.000000     0.000000     0.000000     0.000000     0.000000   
25%      59.000000     0.000000     0.000000     0.000000   220.000000   
50%      68.000000     0.000000   368.500000     0.000000   467.000000   
75%      80.000000   164.000000   733.000000     0.000000   805.500000   
max     313.000000  1600.000000  5644.000000  1526.000000  2336.000000   

       TotalBsmtSF  BsmtFullBath  BsmtHalfBath  GarageYrBlt   GarageCars  \
count  2918.000000   2917.000000   2917.000000  2760.000000  2918.000000   
mean   1051.777587      0.429894      0.061364  1978.113406     1.766621   
std     440.766258      0.524736      0.245687    25.574285     0.761624   
min       0.000000      0.000000      0.000000  1895.000000     0.000000   
25%     793.000000      0.000000      0.000000  1960.000000     1.000000   
50%     989.500000      0.000000      0.000000  1979.000000     2.000000   
75%    1302.000000      1.000000      0.000000  2002.000000     2.000000   
max    6110.000000      3.000000      2.000000  2207.000000     5.000000   

        GarageArea  
count  2918.000000  
mean    472.874572  
std     215.394815  
min       0.000000  
25%     320.000000  
50%     480.000000  
75%     576.000000  
max    1488.000000

#### Data Exploration - contd.
Referring to the details in the <b>Data Description</b> file and comparing with the data types in various feature columns detailed above, there is a mismatch between what the data type is and what it should be. The following changes are made to the feature data types to align them with what is intended by the description:
<ol>
    <li> The datatype of the columns <code>MSSubClass, MSZoning, Street, Alley, LotShape, LandContour, Utilities, LotConfig, LandSlope, Neighborhood, Condition1, Condition2, BldgType, HouseStyle, OverallQual, OverallCond, RoofStyle, RoofMatl, Exterior1st, Exterior2nd, MasVnrType, ExterQual, ExterCond, Foundation, BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1, BsmtFinType2, Heating, HeatingQC, CentralAir, Electrical, KitchenQual, Functional, FireplaceQu, GarageType, GarageFinish, GarageQual, GarageCond, PavedDrive, PoolQC, Fence, MiscFeature, SaleType, SaleCondition</code> should be of type <b>category</b>.
    <li> The datatype of the columns <code>YearBuilt, YearRemodAdd, BsmtFullBath, BsmtHalfBath, FullBath, HalfBath, Bedroom, Kitchen, TotRmsAbvGrd, Fireplaces, GarageYrBlt, GarageCars, MoSold, YrSold</code> should be of type <b>int64</b>.
    <li> The datatype of the columns <code>LotFrontage, LotArea, MasVnrArea, BsmtFinSF1, BsmtFinSF2, BsmtUnfSF, TotalBsmtSF, 1stFlrSF, 2ndFlrSF, LowQualFinSF, GrLivArea, GarageArea, WoodDeckSF, OpenPorchSF, EnclosedPorch, 3SsnPorch, ScreenPorch, PoolArea, MiscVal</code> should be of type <b>float64</b>.
</ol>
For each of these three sets of features we ascertain the details of the missing values.

In [68]:
'''int_64_feats = ['YearBuilt', 'YearRemodAdd', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 'Bedroom', 
                'Kitchen', 'TotRmsAbvGrd', 'Fireplaces', 'GarageYrBlt', 'GarageCars', 'MoSold', 'YrSold']
dat[int_64_feats].isna().sum()

dat[['Kitchen', 'Bedroom']].head()
dat[['KitchenAbvGr', 'BedroomAbvGr']].head()

# It turns out that there is a mismatch between the column names in the Data Description file and the column names
# in the dataset. 'Kitchen' in the Data Description file refers to 'KitchenAbvGr' in the dataset; similarly 
# Bedroom' refers to 'BedroomAbvGr'''

int_64_feats = ['YearBuilt', 'YearRemodAdd', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath', 'HalfBath', 
                'BedroomAbvGr', 'KitchenAbvGr', 'TotRmsAbvGrd', 'Fireplaces', 'GarageYrBlt', 'GarageCars', 
                'MoSold', 'YrSold']
print('\nThe summary of missing values in features that should be of type int64:\n\n',
      dat[int_64_feats].isna().sum())

float_64_feats = ['LotFrontage', 'LotArea', 'MasVnrArea', 'BsmtFinSF1', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 
                  '1stFlrSF', '2ndFlrSF', 'LowQualFinSF', 'GrLivArea', 'GarageArea', 'WoodDeckSF', 'OpenPorchSF', 
                  'EnclosedPorch', '3SsnPorch', 'ScreenPorch', 'PoolArea', 'MiscVal']
print('\nThe summary of missing values in features that should be of type float64:\n\n',
      dat[float_64_feats].isna().sum())

cat_feats = ['MSSubClass', 'MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 
             'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'OverallQual', 
             'OverallCond', 'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType', 'ExterQual', 
             'ExterCond', 'Foundation', 'BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 
             'Heating', 'HeatingQC', 'CentralAir', 'Electrical', 'KitchenQual', 'Functional', 'FireplaceQu', 
             'GarageType', 'GarageFinish', 'GarageQual', 'GarageCond', 'PavedDrive', 'PoolQC', 'Fence', 
             'MiscFeature', 'SaleType', 'SaleCondition']
print('\nThe summary of missing values in features that should be of type category:\n\n',
      dat[cat_feats].isna().sum())



The summary of missing values in features that should be of type int64:

 YearBuilt         0
YearRemodAdd      0
BsmtFullBath      2
BsmtHalfBath      2
FullBath          0
HalfBath          0
BedroomAbvGr      0
KitchenAbvGr      0
TotRmsAbvGrd      0
Fireplaces        0
GarageYrBlt     159
GarageCars        1
MoSold            0
YrSold            0
dtype: int64

The summary of missing values in features that should be of type float64:

 LotFrontage      486
LotArea            0
MasVnrArea        23
BsmtFinSF1         1
BsmtFinSF2         1
BsmtUnfSF          1
TotalBsmtSF        1
1stFlrSF           0
2ndFlrSF           0
LowQualFinSF       0
GrLivArea          0
GarageArea         1
WoodDeckSF         0
OpenPorchSF        0
EnclosedPorch      0
3SsnPorch          0
ScreenPorch        0
PoolArea           0
MiscVal            0
dtype: int64

The summary of missing values in features that should be of type category:

 MSSubClass          0
MSZoning            4
Street              0

From the <b>Data Description</b> file we note that the features <code>Alley, BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1, BsmtFinType2, FireplaceQu, GarageType, GarageFinish, GarageQual, GarageCond, PoolQC, Alley, MiscFeature</code> can take value <b>NA</b> that indicates absence of the relevant feature. However, <b>NA</b> has a specific meaning when used in analysis and hence for all records that report <b>NA</b> in any of the features mentioned above, we would want rewrite <b>NA</b> to <b>ABSNT</b> to better convey the meaning. Similarly, <code>MasVnrType</code> can take value <b>None</b> which we rewrite to <b>ABSNT</b>. Finally, <code>GarageYrBlt</code> can take value <b>NA</b> which we can rewrite to <b>0</b>, if the garage is absent.  

However, care needs to be taken that a genuine <b>NA</b> is not replaced. To that end, we use obvious relationship between features to deduce a genuine <b>NA</b> as follows:
<ul>
    <li> For <code>BsmtQual, BsmtCond, BsmtExposure, BsmtFinType1, BsmtFinType2</code> we ascertain if the corresponding <code>TotalBsmtSF</code> = 0, and if it does then for the respective feature we replace the <b>NA</b> by <b>ABSNT</b>.
    <li> For <code>FireplaceQu</code> we ascertain if the corresponding <code>Fireplaces</code> = 0, and if it does then we replace the <b>NA</b> by <b>ABSNT</b>.
    <li> For <code>GarageType, GarageFinish, GarageQual, GarageCond, GarageYrBlt</code> we ascertain if the corresponding <code>GarageArea</code> = 0, and if it does then for the respective feature we replace the <b>NA</b> by <b>ABSNT</b>.
    <li> For <code>PoolQC</code> we ascertain if the corresponding <code>PoolArea</code> = 0, and if it does then we replace the <b>NA</b> by <b>ABSNT</b>.
    <li> For <code>MiscFeature</code> we ascertain if the corresponding <code>MiscVal</code> = 0, and if it does then we replace the <b>NA</b> by <b>ABSNT</b>.
    <li> For <code>MasVnrType</code> we ascertain if the corresponding <code>MasVnrArea</code> = 0, and if it does then we replace the <b>NA</b> by <b>ABSNT</b>.
    <li> For <code>GarageYrBlt</code> we ascertain if the corresponding <code>GarageArea</code> = 0, and if it does then we replace the <b>NA</b> by <b>0</b>.
    <li> Finally for features, <code>Alley, Fence</code>, because we can't find an obvious relationship with any other features we replace all <b>NA</b> with <b>ABSNT</b>.

We once again, for each of these three sets of features, we ascertain the details of the missing values.

In [81]:
dat = foo

NA_feats = [['BsmtQual', 'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2'],['FireplaceQu'], 
            ['GarageType', 'GarageFinish', 'GarageQual', 'GarageCond', 'GarageYrBlt'], ['PoolQC'], 
            ['MiscFeature'], ['MasVnrType']]
val_feats = ['TotalBsmtSF', 'Fireplaces', 'GarageArea', 'PoolArea', 'MiscVal', 'MasVnrArea']

for i in range(len(NA_feats)):
    for j in range(len(NA_feats[i])):
        #x, y = NA_feats[i][j], val_feats[i]
        dat.loc[(dat.loc[:, NA_feats[i][j]].isna()) & (dat.loc[:, val_feats[i]] == 0), [NA_feats[i][j]]] = 'ABSNT'
        
dat.loc[dat.loc[:, 'Alley'].isna(), 'Alley'] = 'ABSNT'
dat.loc[dat.loc[:, 'Fence'].isna(), 'Fence'] = 'ABSNT'

print('\nThe summary of missing values in features that should be of type int64:\n\n',
      dat[int_64_feats].isna().sum())
print('\nThe summary of missing values in features that should be of type float64:\n\n',
      dat[float_64_feats].isna().sum())
print('\nThe summary of missing values in features that should be of type category:\n\n',
      dat[cat_feats].isna().sum())


The summary of missing values in features that should be of type int64:

 YearBuilt       0
YearRemodAdd    0
BsmtFullBath    2
BsmtHalfBath    2
FullBath        0
HalfBath        0
BedroomAbvGr    0
KitchenAbvGr    0
TotRmsAbvGrd    0
Fireplaces      0
GarageYrBlt     2
GarageCars      1
MoSold          0
YrSold          0
dtype: int64

The summary of missing values in features that should be of type float64:

 LotFrontage      486
LotArea            0
MasVnrArea        23
BsmtFinSF1         1
BsmtFinSF2         1
BsmtUnfSF          1
TotalBsmtSF        1
1stFlrSF           0
2ndFlrSF           0
LowQualFinSF       0
GrLivArea          0
GarageArea         1
WoodDeckSF         0
OpenPorchSF        0
EnclosedPorch      0
3SsnPorch          0
ScreenPorch        0
PoolArea           0
MiscVal            0
dtype: int64

The summary of missing values in features that should be of type category:

 MSSubClass        0
MSZoning          4
Street            0
Alley             0
LotShape     

We now have the better picture of the missing values in the various features. Listing the details of the missing values per datatype per feature:
<ul>
    <ul>Datatype: <b>int64</b>
        <ul>
            <li><code>BsmtFullBath</code> - 2
            <li><code>BsmtHalfBath</code> - 2
            <li><code>GarageYrBlt</code> - 2
            <li><code>GarageCars</code> - 1
         </ul>
    </ul>
    <ul>Datatype: <b>float64</b>
        <ul>
            <li><code>LotFrontage</code> - 486
            <li><code>MasVnrArea</code> - 23
            <li><code>BsmtFinSF1</code> - 1
            <li><code>BsmtFinSF2</code> - 1
            <li><code>BsmtUnfSF</code> - 1
            <li><code>TotalBsmtSF</code> - 1
            <li><code>TotalBsmtSF</code> - 1
            <li><code>GarageArea</code> - 1
         </ul>
    </ul>
    <ul>Datatype: <b>category</b>
        <ul>
            <li><code>MSZoning</code> - 4
            <li><code>Utilities</code> - 2
            <li><code>Exterior1st</code> - 1
            <li><code>Exterior2nd</code> - 1
            <li><code>MasVnrType</code> - 24
            <li><code>BsmtQual</code> - 3
            <li><code>BsmtCond</code> - 4
            <li><code>BsmtExposure</code> - 4
            <li><code>BsmtFinType1</code> - 1
            <li><code>BsmtFinType2</code> - 2
            <li><code>Electrical</code> - 1
            <li><code>KitchenQual</code> - 1
            <li><code>Functional</code> - 2
            <li><code>GarageFinish</code> - 2
            <li><code>GarageQual</code> - 2
            <li><code>GarageCond</code> - 2
            <li><code>PoolQC</code> - 3
            <li><code>MiscFeature</code> - 2
            <li><code>SaleType</code> - 1
         </ul>
     </ul>
</ul>

In [98]:
#dat.loc[dat.loc[:, 'MasVnrType'].isna(), ['Id', 'MasVnrType', 'MasVnrArea']]

dat.loc[:,['YearBuilt', 'YearRemodAdd', 'GarageYrBlt']].head(25)

YearBuilt  YearRemodAdd GarageYrBlt
0        2003          2003        2003
1        1976          1976        1976
2        2001          2002        2001
3        1915          1970        1998
4        2000          2000        2000
5        1993          1995        1993
6        2004          2005        2004
7        1973          1973        1973
8        1931          1950        1931
9        1939          1950        1939
10       1965          1965        1965
11       2005          2006        2005
12       1962          1962        1962
13       2006          2007        2006
14       1960          1960        1960
15       1929          2001        1991
16       1970          1970        1970
17       1967          1967        1967
18       2004          2004        2004
19       1958          1965        1958
20       2005          2006        2005
21       1930          1950        1930
22       2002          2002        2002
23       1976          1976        1976
24       1968          2001        1968

In [102]:
dat.loc[:, ['LotArea', 'LotFrontage']].corr()

LotArea  LotFrontage
LotArea      1.000000     0.489896
LotFrontage  0.489896     1.000000